crates/win-api-wrappers/src/service.rs

```diff
@@ -0,0 +1,142 @@
+
+    fn open_service_with_access(&self, service_name: &str, access: u32) -> Result<Service, Error> {
+        let service_name_wide = WideString::from(service_name);
+
+        // SAFETY: No preconditions.
+        let raw_service_handle = unsafe { OpenServiceW(*self.handle, service_name_wide.as_pcwstr(), access)? };
+
+        // SAFETY: No preconditions.
+        let handle = unsafe { Owned::new(raw_service_handle) };
+
+        Ok(Service { handle })
+    }
+
+    pub fn open_service_read(&self, service_name: &str) -> Result<Service, Error> {
+        self.open_service_with_access(service_name, SERVICE_QUERY_CONFIG | SERVICE_QUERY_STATUS)
+    }
+
+    pub fn open_service_all_access(&self, service_name: &str) -> Result<Service, Error> {
+        self.open_service_with_access(service_name, SERVICE_ALL_ACCESS)
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ServiceStartupMode {
+    Boot,
+    System,
+    Automatic,
+    Manual,
+    Disabled,
+}
+
+pub struct Service {
+    handle: Owned<SC_HANDLE>,
+}
+
+impl Service {
+    pub fn startup_mode(&self) -> Result<ServiceStartupMode, Error> {
+        let mut cbbufsize = 0u32;
+        let mut pcbbytesneeded = 0u32;
+
+        // SAFETY: No preconditions. Query required buffer size.
+        let result = unsafe { QueryServiceConfigW(*self.handle, None, 0, &mut pcbbytesneeded) };
+
+        match result {
+            Err(err) if err.code() == ERROR_INSUFFICIENT_BUFFER.to_hresult() => {
+                // Expected error, continue.
+            }
+            Err(err) => return Err(err.into()),
+            Ok(_) => panic!("QueryServiceConfigW should fail with ERROR_INSUFFICIENT_BUFFER"),
+        }
+
+        // Event though `QueryServiceConfigW` states that `lpserviceconfig` is a `buffer`,
+        // it needs to be aligned ir order to avoid undefined behavior when accessing
+        // `QUERY_SERVICE_CONFIGW.
+        //
+        // Clippy warning:
+        // casting from `*mut u8` to a more-strictly-aligned pointer
+        // (`*mut windows::Win32::System::Services::QUERY_SERVICE_CONFIGW`) (1 < 8 bytes)
+        assert_eq!(align_of::<QUERY_SERVICE_CONFIGW>(), size_of::<u64>());
+
+        let pcbbytesneeded_usize = usize::try_from(pcbbytesneeded).expect("pcbbytesneeded < 8K as per MSDN");
+
+        let aligned_u64_array_size =
+            pcbbytesneeded_usize / size_of::<u64>() + usize::from(pcbbytesneeded_usize % size_of::<u64>() != 0);
+
+        let aligned_size = u32::try_from(aligned_u64_array_size * size_of::<u64>()).expect("pcbbytesneeded <= 8K");
+
+        let mut buffer: Box<Vec<u64>> = Box::new(vec![0; aligned_u64_array_size]);
+
+        // SAFETY: Buffuer with correct size was allocated prior to the call.
+        unsafe {
+            QueryServiceConfigW(
+                *self.handle,
+                Some(buffer.as_mut_ptr() as *mut QUERY_SERVICE_CONFIGW),
+                aligned_size,
+                &mut cbbufsize,
+            )?
+        };
```

Comment:
suggestion: What about using `RawBuffer`?

https://github.com/Devolutions/devolutions-gateway/blob/ebc6ba46ba6009193dd0987e7281311398e5388f/crates/win-api-wrappers/src/raw_buffer.rs#L27

I feel your code is kind of re-implementing a similar logic in an indirect way using `Vec` and `u64`.

See an example usage here:

https://github.com/Devolutions/devolutions-gateway/blob/ebc6ba46ba6009193dd0987e7281311398e5388f/crates/win-api-wrappers/src/token.rs#L232-L276

BUT!

If the struct is not variable-sized (DST), the best is to initialize it on the stack using `let mut config = QUERY_SERVICE_CONFIGW::default()` and pass it as a mut reference `&mut config`. If necessary, convert into a `*mut c_void` pointer (`&mut config as *mut c_void`).

Reply:
Indeed! `RawBuffer` is what I need, thanks!

Reply:
NOTE: I still needed to silence Clippy (`windows` crate still takes buffer via `*mut QUERY_SERVICE_CONFIGW` and  *mut u8 -> *mut QUERY_SERVICE_CONFIGW` cast still should occur), but now thanks to RawBuffer it is guaranteed to be aligned.